File: sql-insight-engine/sql-insight-engine-innodb/src/main/java/org/gongxuanzhang/sql/insight/core/engine/storage/innodb/core/InnodbBTree.java
/*
 * Copyright 2023 java-mysql  and the original author or authors <gongxuanzhangmelt@gmail.com>.
 *
 * Licensed under the GNU Affero General Public License v3.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://github.com/implement-study/sql-insight/blob/main/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gongxuanzhang.sql.insight.core.engine.storage.innodb.core;

import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.page.RootPage;
import org.gongxuanzhang.sql.insight.core.engine.storage.innodb.utils.PageSupport;
import org.gongxuanzhang.sql.insight.core.object.Index;
import org.gongxuanzhang.sql.insight.core.object.InsertRow;

/**
 * innodb b+ tree
 *
 * @author gongxuanzhangmelt@gmail.com
 **/
public interface InnodbBTree extends Index {

    /**
     * get the b+ trees root node.
     * page is node
     *
     * @return not null
     **/
    default RootPage getIndexRoot() {
        return PageSupport.getRoot(this);
    }

    /**
     * insert a row.
     **/
    @Override
    void insert(InsertRow row);


}

File: sql-insight-engine/sql-insight-engine-innodb/src/main/kotlin/tech/insight/engine/innodb/core/InnodbBTree.kt
package tech.insight.engine.innodb.core

import tech.insight.core.bean.Index
import tech.insight.core.bean.InsertRow
import tech.insight.core.bean.Row
import tech.insight.engine.innodb.page.IndexKey
import tech.insight.engine.innodb.page.InnoDbPage

/**
 * innodb b+ tree
 *
 * @author gongxuanzhangmelt@gmail.com
 */
interface InnodbBTree : Index {
    /**
     * insert a row.
     */
    override fun insert(row: InsertRow)


    /**
     * b tree can find the row
     *
     */
    fun findByKey(key: IndexKey): Row
}
